### PR TITLE
TINY-9157: Move TestStore out of alloy and into agar

### PR DESCRIPTION
**Pull request by shanmen-tiny**
Related Ticket: TINY-9157

Description of Changes:
* Moved TestStore out of alloy and into agar

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] ~Docs ticket created (if applicable)~

GitHub issues (if applicable):

### CODE REVIEW

File: modules/alloy/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## Unreleased
 
+### Removed
+- Removed `TestStore` from `TestHelpers` export, moved `TestStore` to Agar. #TINY-9157
```
Comment (tiny-ben-tran) [inline modules/alloy/CHANGELOG.md:10]:
```suggestion
- Moved `TestStore` to Agar. #TINY-9157
```

File: modules/agar/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## Unreleased
 
+### Added
+- Added `TestStore` API, which is moved from Alloy. #TINY-9157
```
Comment (tiny-ben-tran) [inline modules/agar/CHANGELOG.md:10]:
```suggestion
- Added `TestStore` API, which was moved from Alloy. #TINY-9157
```

**Review by tiny-ben-tran — APPROVED** [review]

**Review by ephox-mogran — COMMENTED** [review]
Looks good. The one thing that I didn't add to the ticket, but was in one of the many slack conversations, was we did want to make one change to TestStore.

Currently, there is a function called this: 

```
const sAssertSortedEq = (label: string, expected: any[]) => Step.sync(() =>
  // Can't use a normal step here, because we don't need to get array lazily
    Assert.eq(label, expected.slice(0).sort(), array.slice(0).sort())
  );
```

We wanted to add an additional non-Step equivalent: `assertSortedEq`. 

- [ ] add `assertSortedEq` to TestStore in agar.

File: modules/agar/src/main/ts/ephox/agar/api/TestStore.ts [hunk 3]
```diff
@@ -55,6 +58,8 @@ export const TestStore = (): TestStore => {
 
   const assertEq = (label: string, expected: any[]) => Assert.eq(label, expected, array.slice(0));
 
+  const assertSortedEq = (label: string, expected: any[]) => Assert.eq(label, expected.slice(0).sort(), array.slice(0).sort());
+
   const sAssertSortedEq = (label: string, expected: any[]) => Step.sync(() =>
   // Can't use a normal step here, because we don't need to get array lazily
```
Comment (ephox-mogran) [inline modules/agar/src/main/ts/ephox/agar/api/TestStore.ts:64]: We should probably get rid of that comment. I can't work out what it is intended to mean. I think there's a typo in there somewhere.

**Review by ephox-mogran — COMMENTED** [review]
I'm not sure if the `develop` is the branch you should be targeting with the PR. I'll confirm and get back to you.

- [x] confirm target branch

File: modules/agar/src/main/ts/ephox/agar/api/TestStore.ts [hunk 4]
```diff
@@ -55,6 +58,8 @@ export const TestStore = (): TestStore => {
...
     Assert.eq(label, expected.slice(0).sort(), array.slice(0).sort())
```
Comment (lnewson) [inline modules/agar/src/main/ts/ephox/agar/api/TestStore.ts:65]: This also should not have duplicate code. You now have 2 implementations for `assertSortedEq` and they should be using the same implementation.

- [ ] Make `sAssertSortedEq` call `assertSortedEq` to remove duplication

File: modules/alloy/src/main/ts/ephox/alloy/api/testhelpers/TestHelpers.ts [hunk 5]
```diff
@@ -1,8 +1,6 @@
 // Test code - should eventually move to a separate project
 import * as GuiSetup from './GuiSetup';
-import { TestStore } from './TestStore';
 
 export {
-  GuiSetup,
-  TestStore
```
Comment (lnewson) [inline modules/alloy/src/main/ts/ephox/alloy/api/testhelpers/TestHelpers.ts:7]: If you're removing this you need to log a Jira to upgrade all the premium plugins using it as well.

- [ ] Ensure followup Jira exists to upgrade anything else that uses `TestStore`.

File: modules/agar/src/main/ts/ephox/agar/api/TestStore.ts (hunk 3 above)
Comment (shanmen-tiny) [reply modules/agar/src/main/ts/ephox/agar/api/TestStore.ts]: Removed the comment.

File: modules/agar/src/main/ts/ephox/agar/api/TestStore.ts (hunk 4 above)
Comment (shanmen-tiny) [reply modules/agar/src/main/ts/ephox/agar/api/TestStore.ts]: Removed duplicated call in `sAssertSortedEq`, `sAssertSortedEq` is now calling `assertSortedEq`

File: modules/alloy/src/main/ts/ephox/alloy/api/testhelpers/TestHelpers.ts (hunk 5 above)
Comment (shanmen-tiny) [reply modules/alloy/src/main/ts/ephox/alloy/api/testhelpers/TestHelpers.ts:7]: Logged a ticket TINY-9166 to follow up and upgrading the premium plugins that uses TestStore.

**Review by ephox-mogran — APPROVED** [review]

File: modules/agar/src/main/ts/ephox/agar/api/TestStore.ts [hunk 6]
```diff
@@ -11,7 +13,8 @@ export interface TestStore {
   cClear: Chain<any, any>;
   sAssertEq: <T> (label: string, expected: any[]) => Step<T, T>;
   cAssertEq: <T> (label: string, expected: any[]) => Chain<T, T>;
-  assertEq: (label: string, expected: any[]) => void;
+  assertEq: (labely: string, expected: any[]) => void;
+  assertSortedEq: (labely: string, expected: any[]) => void;
```
Comment (spocke) [inline modules/agar/src/main/ts/ephox/agar/api/TestStore.ts:17]: It's a bit weird having the types by any then have sorting on this since the sort function will not sort non primitive objects unless you have a sort function. Also why is this entire thing any wouldn't a type variable be useful here the expected would be an array of what you add?

- [ ] Consider using type variables instead of just adding a new module that allows any all over and has no type safety
- [ ] Consider adding an explicit sort comparator function so the type signature would be `assertSortedEq(label: string, expected: T[], cmp: (a: T, b: T): number): void;`

**Review by spocke — APPROVED** [review]

File: modules/agar/src/main/ts/ephox/agar/api/TestStore.ts (hunk 6 above)
Comment (shanmen-tiny) [reply modules/agar/src/main/ts/ephox/agar/api/TestStore.ts]: Added types and comparator function.

File: modules/agar/src/test/ts/atomic/api/TestStoreTest.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+import * as GeneralSteps from 'ephox/agar/api/GeneralSteps';
+import { Pipeline } from 'ephox/agar/api/Pipeline';
+import { Step } from 'ephox/agar/api/Step';
+import { TestStore } from 'ephox/agar/api/TestStore';
+
+interface TestData {
+  name: string;
+  age: number;
+  isHappy: boolean;
```
Comment (lnewson) [inline modules/agar/src/test/ts/atomic/api/TestStoreTest.ts:12]: Readonly please.

File: modules/agar/src/test/ts/atomic/api/TestStoreTest.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+UnitTest.asyncTest('TestStoreTest strings', (success, failure) => {
```
Comment (lnewson) [inline modules/agar/src/test/ts/atomic/api/TestStoreTest.ts:15]: New tests should be using `describe` and `it` BDD style instead of the deprecated/legacy `UnitTest` module.

- [ ] Swap to using BDD style testing

File: modules/alloy/src/test/ts/browser/behaviour/dragging/MouseDragEventTest.ts [hunk 9]
```diff
@@ -9,8 +9,13 @@ import * as GuiFactory from 'ephox/alloy/api/component/GuiFactory';
 import * as GuiSetup from 'ephox/alloy/api/testhelpers/GuiSetup';
 import { Container } from 'ephox/alloy/api/ui/Container';
 
+interface StoreMouseDragEventTest {
+  left: number;
+  top: number;
```
Comment (lnewson) [inline modules/alloy/src/test/ts/browser/behaviour/dragging/MouseDragEventTest.ts:14]: Readonly please.

File: modules/alloy/src/test/ts/browser/behaviour/dragndrop/DragnDropTest.ts [hunk 10]
```diff
@@ -9,8 +9,14 @@ import * as GuiSetup from 'ephox/alloy/api/testhelpers/GuiSetup';
 import { Container } from 'ephox/alloy/api/ui/Container';
 import { DropDragndropConfigSpec, StartingDragndropConfigSpec } from 'ephox/alloy/dragging/dragndrop/DragnDropTypes';
 
+interface StoreDragnDropTest {
+  type: string;
+  files: [] | Array<Record<string, any>>;
+  data?: string | Record<string, any>;
```
Comment (lnewson) [inline modules/alloy/src/test/ts/browser/behaviour/dragndrop/DragnDropTest.ts:15]: More readonly here.

File: modules/alloy/src/test/ts/browser/events/GuiEventsTest.ts [hunk 11]
```diff
@@ -4,6 +4,11 @@ import { Attribute, DomEvent, EventArgs, Insert, Remove, SugarElement, SugarNode
 
 import * as GuiEvents from 'ephox/alloy/events/GuiEvents';
 
+interface StoreGuiEventsTest {
+  eventName: string;
+  target?: string;
```
Comment (lnewson) [inline modules/alloy/src/test/ts/browser/events/GuiEventsTest.ts:9]: Readonly.

File: modules/alloy/src/main/ts/ephox/alloy/api/testhelpers/TestHelpers.ts (hunk 5 above)
Comment (lnewson) [reply modules/alloy/src/main/ts/ephox/alloy/api/testhelpers/TestHelpers.ts:7]: This Jira has no details... please fill it out and ensure it has the issue and what needs to be done. It also shouldn't be assigned to the Editor Team as it could be done by anyone.

- [ ] Fix Jira

**Review by lnewson — CHANGES_REQUESTED** [review]
Looks pretty good, but there's some smaller issues with the recent changes.

File: modules/alloy/src/test/ts/browser/behaviour/PinchingTest.ts [hunk 12]
```diff
@@ -8,9 +8,15 @@ import * as AlloyTriggers from 'ephox/alloy/api/events/AlloyTriggers';
 import * as NativeEvents from 'ephox/alloy/api/events/NativeEvents';
 import * as GuiSetup from 'ephox/alloy/api/testhelpers/GuiSetup';
 
+interface StorePinchingTest {
+  method: string;
+  dx: number;
+  dy: number;
```
Comment (lnewson) [inline modules/alloy/src/test/ts/browser/behaviour/PinchingTest.ts:14]: Readonly here as well sorry.

File: modules/agar/src/test/ts/atomic/api/TestStoreTest.ts (hunk 8 above)
Comment (shanmen-tiny) [reply modules/agar/src/test/ts/atomic/api/TestStoreTest.ts]: Sorry, I copied the other tests in agar which is using the legacy `UnitTest` module. Have converted them to using `describe` and `it`.

File: modules/alloy/src/main/ts/ephox/alloy/api/testhelpers/TestHelpers.ts (hunk 5 above)
Comment (shanmen-tiny) [reply modules/alloy/src/main/ts/ephox/alloy/api/testhelpers/TestHelpers.ts:7]: I have added more details and DoD to the ticket TINY-9166, hope it makes more sense. I have also removed the assigned team.

File: modules/agar/src/test/ts/atomic/api/TestStoreTest.ts [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+import { TestStore } from 'ephox/agar/api/TestStore';
+
+interface TestData {
+  readonly name: string;
+  readonly age: number;
+  readonly isHappy: boolean;
+}
+
+describe('atomic.agar.api.TestStoreTest', () => {
+  it('TestStore strings', () => {
```
Comment (ephox-mogran) [inline modules/agar/src/test/ts/atomic/api/TestStoreTest.ts:16]: Nice tests. Maybe spilt the Step tests and the non-Step tests into different `context` definitions, to make the separation a bit more obvious.

**Review by ephox-mogran — APPROVED** [review]
Nice work @shanmen-tiny. Adding the types that @spocke suggested really helped identify just what we were testing in some of those tests, so I think this change has improved the testing side of the codebase (at least in terms of maintainability).

File: modules/agar/CHANGELOG.md [hunk 14]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## Unreleased
 
+### Added
+- Added `TestStore` API, which was moved from Alloy. #TINY-9157
```
Comment (lnewson) [inline modules/agar/CHANGELOG.md:10]: I could have sworn I left a suggestion for this but I can't find it sorry. Anyways, maybe something like this to clarify it also needs types.

```suggestion
- Added `TestStore` API, which was moved from Alloy and requires types. #TINY-9157
```

File: modules/agar/src/main/ts/ephox/agar/api/TestStore.ts [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  sAssertSortedEq: (label: string, expected: T[], cmp?: (a: T, b: T) => number) => Step<any, any>;
+}
+
+export const TestStore = <T = string>(): TestStore<T> => {
+  let array: T[] = [];
+
+  const add = (value: T) => {
+    array.push(value);
+    // eslint-disable-next-line no-console
+    console.log('store.add', value, array);
```
Comment (lnewson) [inline modules/agar/src/main/ts/ephox/agar/api/TestStore.ts:27]: Not exactly something to fix here, but we don't include console logs in tests. It's another days discussion though given it used to do this.

File: modules/agar/src/test/ts/atomic/api/TestStoreTest.ts [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    { name: 'CCCCCCC', age: 999, isHappy: false },
+    { name: 'MMM', age: 20, isHappy: true }
+  ];
+
+  const objectsSortedByName = Arr.sort(objects, (a, b) => a.name.length - b.name.length);
+  const objectsSortedByAge = Arr.sort(objects, (a, b) => a.age - b.age);
+  const objectsSortedByIsHappy = Arr.sort(objects, (a, b) => Number(a.isHappy) - Number(b.isHappy));
+
+  context('Step based', () => {
+    it('TestStore strings', () => {
```
Comment (lnewson) [inline modules/agar/src/test/ts/atomic/api/TestStoreTest.ts:33]: We should include Jira numbers as prefixes on strings so it's easy to see why something was added. This is a little bit of an edge case since it was just moved though. Up to you all if you include them on the `it` blocks.

**Review by lnewson — APPROVED** [review]

File: modules/agar/src/test/ts/atomic/api/TestStoreTest.ts (hunk 16 above)
Comment (shanmen-tiny) [reply modules/agar/src/test/ts/atomic/api/TestStoreTest.ts]: Thanks Lee. Happy to include the ticket numbers as prefixes on `it` blocks to make things clearer.
